### ISSUE TITLE
Needs to embed a fake struct to serialize direct STL types

### ISSUE THREAD

**Issue by krawq** [p-ranav/alpaca#14]
This works:
`constexpr auto OPTIONS = alpaca::options::with_checksum | alpaca::options::fixed_length_encoding;
using serialize_t = std::map<int,bool>;
auto Serialize(const serialize_t& to_serialize)
{
	std::vector<uint8_t> buffer;
	struct MyStruct {
			  serialize_t field;
			};
	const MyStruct& object{to_serialize};
	if(auto bytes_written=alpaca::serialize<OPTIONS>(object,buffer))
		return std::make_pair(buffer,true);
	else
		return std::make_pair(buffer,false);
};`

This does not:

`constexpr auto OPTIONS = alpaca::options::with_checksum | alpaca::options::fixed_length_encoding;
using serialize_t = std::map<int,bool>;
auto Serialize(const serialize_t& to_serialize)
{
	std::vector<uint8_t> buffer;
	if(auto bytes_written=alpaca::serialize<OPTIONS>(to_serialize,buffer))
		return std::make_pair(buffer,true);
	else
		return std::make_pair(buffer,false);
};`

Would be great to be able to use it without using MyStruct.

**Comment by krawq** [comment]
@p-ranav Any comment ?

**Comment by p-ranav** [comment]
There are internal functions that can be used for this. Perhaps these can be moved out of detail and updated to match the API interface. 

The function you're looking for is `to_bytes(...)`

```cpp
// main.cpp
#include <alpaca/alpaca.h>

constexpr auto OPTIONS =
    alpaca::options::with_checksum | alpaca::options::fixed_length_encoding;

using serialize_t = std::map<int, bool>;

auto Serialize(const serialize_t &to_serialize) {
  std::vector<uint8_t> buffer;
  std::size_t byte_index{0};
  alpaca::detail::to_bytes<OPTIONS>(buffer, byte_index, to_serialize);
  return buffer;
};

int main() {
  auto foo = serialize_t{{5, true}, {3, false}};
  auto buffer = Serialize(foo);

  alpaca::detail::print_bytes(buffer);
}
```

```console
foo:bar $ ./main 
bytes[18]:
  0x02 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  0x03 0x00 0x00 0x00 0x00 0x05 0x00 0x00 
  0x00 0x01 

```

**Comment by p-ranav** [comment]
Yes, I think so. We can add overloads to `to_bytes` that match the serialize interface (`serialize(object, bytes)`). The call can be routed to the relevant `to_bytes` implementation for that type. Just need to make sure that the OPTIONS still work correctly - calculating checksum if requested, calculating a type-hash if requested etc.
